### PR TITLE
DIY Loop dev sync

### PR DESCRIPTION
**Pull request by ps2**

### CODE REVIEW

File: Loop/de.lproj/InfoPlist.strings [hunk 1]
```diff
@@ -8,11 +8,11 @@
 "NSBluetoothPeripheralUsageDescription" = "Bluetooth wird verwendet, um mit Insulinpumpen und kontinuierlichen Blutzuckermessgeräten zu kommunizieren";
 
 /* Privacy - Face ID Usage Description */
-"NSFaceIDUsageDescription" = "Face ID wird benutzt um Insulin-Bolus zu authentifizieren.";
```
Comment (darinkrauss) [inline Loop/de.lproj/InfoPlist.strings:11]: If these localization changes are specific to German, we might want to pull them into a separate PR next time to address the specific language "bugs". Fine for now, though.

Comment (darinkrauss) [reply Loop/de.lproj/InfoPlist.strings:11]: This actually brings up a good point about code reviews. How do I review language changes for a language I don't understand?

File: WatchApp Extension/Models/GlucoseChartData.swift [hunk 2]
```diff
@@ -95,9 +95,9 @@ private extension HKUnit {
 
     var lowWatermark: Double {
         if self == .milligramsPerDeciliter {
-            return 50.0
+            return 75
```
Comment (darinkrauss) [inline WatchApp Extension/Models/GlucoseChartData.swift:98]: Shouldn't this be in sync with the mmol/L value? Also, shouldn't it be configurable?

File: WatchApp/Base.lproj/Interface.storyboard [hunk 3]
```diff
@@ -23,8 +24,8 @@
                                                 <group width="22" height="22" alignment="left" verticalAlignment="center" radius="11" id="J90-Xm-Znp">
                                                     <items>
                                                         <button width="22" height="22" alignment="left" verticalAlignment="center" accessibilityLabel="Subtract" title="−" id="Dh9-HV-fXy">
-                                                            <color key="titleColor" red="0.38823529410000002" green="0.85490196080000003" blue="0.21960784310000001" alpha="1" colorSpace="custom" customColorSpace="sRGB"/>
-                                                            <color key="backgroundColor" red="0.38823529410000002" green="0.85490196080000003" blue="0.21960784310000001" alpha="0.14000000000000001" colorSpace="custom" customColorSpace="sRGB"/>
+                                                            <color key="titleColor" name="carbs"/>
+                                                            <color key="backgroundColor" name="carbs-dark"/>
```
Comment (darinkrauss) [inline WatchApp/Base.lproj/Interface.storyboard:28]: Note the background color has gone from an alpha of 0.140 to 1.000 (and RGBs adjusted accordingly). This works fine for a white background, but not otherwise. Just wanted to check to see if this was intended.

File: WatchApp/Base.lproj/Interface.storyboard [hunk 4]
```diff
@@ -33,14 +34,14 @@
...
+                                                    <color key="textColor" name="carbs"/>
                                                     <fontDescription key="font" type="system" pointSize="40"/>
                                                 </label>
                                                 <group width="22" height="22" alignment="right" verticalAlignment="center" radius="11" id="6ip-X6-F2J">
                                                     <items>
                                                         <button width="22" height="22" alignment="right" verticalAlignment="center" accessibilityLabel="Add" title="+" id="eu3-pj-GH3">
-                                                            <color key="titleColor" red="0.38823529410000002" green="0.85490196080000003" blue="0.21960784310000001" alpha="1" colorSpace="custom" customColorSpace="sRGB"/>
-                                                            <color key="backgroundColor" red="0.38823529410000002" green="0.85490196080000003" blue="0.21960784310000001" alpha="0.14000000000000001" colorSpace="custom" customColorSpace="sRGB"/>
```
Comment (darinkrauss) [inline WatchApp/Base.lproj/Interface.storyboard:43]: Same here.

Comment (darinkrauss) [reply WatchApp/Base.lproj/Interface.storyboard:43]: And others below, but I'm not going to comment on them all.

Comment (darinkrauss) [reply WatchApp/Base.lproj/Interface.storyboard:43]: Looks like `carbs-dark`, `glucose-dark`, and `workout-dark` all had their alpha channel set to 1.000 (and RGB values adjusted).

File: WatchApp/Base.lproj/Interface.storyboard [hunk 5]
```diff
@@ -503,5 +543,31 @@
             <point key="canvasLocation" x="38" y="1153"/>
         </scene>
     </scenes>
+    <resources>
```
Comment (darinkrauss) [inline WatchApp/Base.lproj/Interface.storyboard:546]: I notice all of these same colors in the Assets. Is this duplicate or just Xcode cached data?

File: WatchApp Extension/Scenes/GlucoseChartValueHashable.swift [hunk 6]
```diff
@@ -78,10 +78,10 @@ extension AbsoluteScheduleValue: GlucoseChartValueHashable where T == Range<HKQu
 
 extension GlucoseRangeSchedule.Override: GlucoseChartValueHashable {
     var min: Double {
-        return value.minValue
+        return quantityRange.lowerBound.doubleValue(for: .milligramsPerDeciliter)
```
Comment (darinkrauss) [inline WatchApp Extension/Scenes/GlucoseChartValueHashable.swift:81]: Should this (and 4 lines below) use `unit`?

File: WatchApp Extension/Managers/LoopDataManager.swift [hunk 7]
```diff
@@ -48,9 +51,18 @@ class LoopDataManager {
...
+            cacheStore: cacheStore,
+            defaultAbsorptionTimes: LoopSettings.defaultCarbAbsorptionTimes,
+            syncVersion: 0
+        )
         glucoseStore = GlucoseStore(
-            healthStore: HKHealthStore(),
-            cacheStore: PersistenceController.controllerInLocalDirectory(),
+            healthStore: healthStore,
+            cacheStore: cacheStore,
             cacheLength: .hours(4)
```
Comment (darinkrauss) [inline WatchApp Extension/Managers/LoopDataManager.swift:66]: Only 4 hours? Constant or configurable?

File: LoopCore/CarbEntryUserInfo.swift [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+
+extension CarbEntryUserInfo: RawRepresentable {
+    public typealias RawValue = [String: Any]
+
+    static let version = 2
+    public static let name = "CarbEntryUserInfo"
+
+    public init?(rawValue: RawValue) {
+        guard rawValue["v"] as? Int == type(of: self).version && rawValue["name"] as? String == CarbEntryUserInfo.name,
```
Comment (darinkrauss) [inline LoopCore/CarbEntryUserInfo.swift:29]: Do we need to support the version 1 of this struct? (Perhaps a noob question because I don't entirely understand when/how the raw values are serialized/deserialized yet.)

File: Common/Extensions/NSTimeInterval.swift [hunk 9]
```diff
@@ -30,6 +34,10 @@ extension TimeInterval {
         self.init(minutes: hours * 60)
     }
 
+    init(days: Double) {
+        self.init(hours: days * 24)
+    }
+
     var minutes: Double {
         return self / 60.0
     }
```
Comment (darinkrauss) [inline Common/Extensions/NSTimeInterval.swift]: Do we want a `days` property (much like `minutes` or `hours`) to match the initializers?

File: Loop/View Controllers/ChartsTableViewController.swift [hunk 10]
```diff
@@ -168,8 +166,11 @@ class ChartsTableViewController: UITableViewController, UIGestureRecognizerDeleg
...
 
-    var active: Bool = false {
-        didSet {
+    var active: Bool {
+        get {
+            return UIApplication.shared.applicationState == .active
+        }
+        set {
             log.debug("[reloadData] for app change to active: %d", active)
             reloadData()
```
Comment (darinkrauss) [inline Loop/View Controllers/ChartsTableViewController.swift:175]: Do we want to reload the data when it is changing to *not* active?

File: WatchApp Extension/Controllers/HUDRowController.swift [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+    @IBOutlet private var outerGroup: WKInterfaceGroup!
+    @IBOutlet private var bottomSeparator: WKInterfaceSeparator!
+}
+
+extension HUDRowController {
+    func setTitle(_ title: String) {
+        textLabel.setText(title.localizedUppercase)
+    }
+
+    func setContentInset(_ inset: NSDirectionalEdgeInsets) {
```
Comment (darinkrauss) [inline WatchApp Extension/Controllers/HUDRowController.swift]: [nit] Order of funcs to match order of outlets above.

File: Learn/Extensions/NSNumber.swift [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+
+
+extension NSNumber {
+    convenience init?(value: Double?) {
+        if let value = value {
+            self.init(value: value)
+        } else {
+            return nil
+        }
+    }
```
Comment (mpangburn) [inline Learn/Extensions/NSNumber.swift]: Seems clearer to me to unwrap the Double argument at the call site rather than overloading `init` to fail if its argument is `nil`.
If the extra unwrap is undesirable, the generic way to apply a function "inside" an Optional is `map`, i.e. `value.map(NSNumber.init(value:))`.

File: LoopUI/Managers/StatusChartsManager.swift [hunk 13]
```diff
@@ -951,16 +952,9 @@ public final class StatusChartsManager {
...
-            if let override = schedule.override {
-                targetOverridePoints = ChartPoint.pointsForGlucoseRangeScheduleOverride(override, unit: schedule.unit, xAxisValues: xAxisValues, extendEndDateToChart: true)
-
-                targetOverrideDurationPoints = ChartPoint.pointsForGlucoseRangeScheduleOverride(override, unit: schedule.unit, xAxisValues: xAxisValues)
-            } else {
-                targetOverridePoints = []
-                targetOverrideDurationPoints = []
-            }
+            targetGlucosePoints = ChartPoint.pointsForGlucoseRangeSchedule(schedule, unit: glucoseUnit, xAxisValues: xAxisValues)
+            targetOverridePoints = ChartPoint.pointsForGlucoseRangeScheduleOverride(schedule, unit: glucoseUnit, xAxisValues: xAxisValues, extendEndDateToChart: true)
```
Comment (darinkrauss) [inline LoopUI/Managers/StatusChartsManager.swift:956]: [noob] We appear to have lost the `schedule.override` aspect here. Not sure if that means anything, but thought I'd call it out.

File: Learn/Lessons/TimeInRangeLesson.swift [hunk 14]
```diff
@@ -0,0 +1,265 @@
...
+                    )
+                ])
+            }
+        }
+    }
+}
+
+
+private extension QuantityRangeEntry {
+    class func glucoseRange(quantityFormatter: QuantityFormatter, unit: HKUnit) -> QuantityRangeEntry {
```
Comment (mpangburn) [inline Learn/Lessons/TimeInRangeLesson.swift]: A glucose range entry is probably useful as a reusable component; worth making `internal` and moving to QuantityRangeEntry.swift?

File: Learn/Lesson.swift [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+    var configurationSections: [LessonSectionProviding] { get }
+
+    func execute(completion: @escaping (_ resultSections: [LessonSectionProviding]) -> Void)
+}
+
+
+protocol LessonSectionProviding {
+    var headerTitle: String? { get }
+
+    var footerTitle: String? { get }
```
Comment (mpangburn) [inline Learn/Lesson.swift:28]: Several LessonSectionProviding implementations have no header or footer title. Might be nice to default these properties to `nil` in an extension.

File: Learn/Lessons/TimeInRangeLesson.swift [hunk 16]
```diff
@@ -0,0 +1,265 @@
...
+
+        return sum / Double(count)
+    }
+
+    var dictionary: [DateInterval: Double] {
+        guard let allDates = allDates, let averagePercentInRange = averagePercentInRange else {
+            return [:]
+        }
+
+        return [allDates: averagePercentInRange]
```
Comment (mpangburn) [inline Learn/Lessons/TimeInRangeLesson.swift]: Without looking at call sites, it's not obvious why a `dictionary` property is desirable here, since it only ever has zero or one key-value pairs. A named, optional tuple makes more sense to me.

File: Learn/Lessons/TimeInRangeLesson.swift [hunk 17]
```diff
@@ -0,0 +1,265 @@
...
+                    for sample in samples {
+                        if self.range.contains(sample.quantity) {
+                            inRangeCount += 1
+                        }
+                        totalCount += 1
+                    }
+
+                    if totalCount > 0 {
+                        _ = lockedResults.mutate({ (results) in
+                            results[interval] = Double(inRangeCount) / Double(totalCount)
```
Comment (mpangburn) [inline Learn/Lessons/TimeInRangeLesson.swift]: The logic for computing the time in range is a great fit for a generic extension:
```swift
extension Sequence {
    func proportion(where isIncluded: (Element) -> Bool) -> Double? {
        return average(by: { isIncluded($0) ? 1 : 0 })
    }

    func average<T: FloatingPoint>(by getMetric: (Element) -> T) -> T? {
        let (sum, count) = reduce(into: (sum: 0 as T, count: 0)) { result, element in
            result.0 += getMetric(element)
            result.1 += 1
        }

        guard count > 0 else {
            return nil
        }

        return sum / T(count)
    }
}

extension Sequence where Element: FloatingPoint {
    func average() -> Element? {
        return average(by: { $0 })
    }
}
```

The call site here becomes beautiful:
```swift
let timeInRange = samples.proportion(where: { self.range.contains($0.quantity) }) 
```

File: Learn/View Controllers/LessonConfigurationViewController.swift [hunk 18]
```diff
@@ -0,0 +1,161 @@
...
+        }
+    }
+}
+
+
+extension LessonConfigurationViewController: DatePickerTableViewCellDelegate {
+    func datePickerTableViewCellDidUpdateDate(_ cell: DatePickerTableViewCell) {
+
+    }
+}
```
Comment (mpangburn) [inline Learn/View Controllers/LessonConfigurationViewController.swift:161]: The empty body of this method is perhaps an indicator that `hideDatePickerCells(excluding:)` doesn't need to be constrained to DatePickerTableViewCellDelegate implementors.
(If I were reading this and were unaware of the declaration of that function, I would've been confused by this protocol conformance, since the VC is never actually used as a delegate.)

File: LoopCore/NSUserDefaults.swift [hunk 19]
```diff
@@ -21,7 +21,9 @@ extension UserDefaults {
         case insulinSensitivitySchedule = "com.loudnate.Naterade.InsulinSensitivitySchedule"
     }
 
-    var basalRateSchedule: BasalRateSchedule? {
+    public static let appGroup = UserDefaults(suiteName: Bundle.main.appGroupSuiteName)
```
Comment (mpangburn) [inline LoopCore/NSUserDefaults.swift:24]: This property is now an Optional when it wasn't before—is there a known occasion when we expect to have `nil` here?

File: LoopCore/PersistenceController.swift [hunk 20]
```diff
@@ -9,19 +9,19 @@ import LoopKit
...
-    class func controllerInAppGroupDirectory() -> PersistenceController {
+    public class func controllerInAppGroupDirectory(isReadOnly: Bool = false) -> PersistenceController {
         let appGroup = Bundle.main.appGroupSuiteName
         guard let directoryURL = FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: appGroup) else {
             assertionFailure("Could not get a container directory URL. Please ensure App Groups are set up correctly in entitlements.")
             return self.init(directoryURL: URL(fileURLWithPath: "/"))
         }
 
-        let isReadOnly = Bundle.main.bundleURL.pathExtension == "appex"
+        let isReadOnly = isReadOnly || Bundle.main.bundleURL.pathExtension == "appex"
```
Comment (mpangburn) [inline LoopCore/PersistenceController.swift]: An `isAppExtension` computed variable on Bundle might be nice here.

File: WatchApp Extension/Controllers/CarbEntryListController.swift [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+    private func updateActiveCarbs() {
+        guard let activeCarbohydrates = loopManager.activeContext?.activeCarbohydrates else {
+            return
+        }
+
+        cobLabel.setText(carbFormatter.string(from: activeCarbohydrates, for: .gram()))
+    }
+
+    private func reloadCarbEntries() {
+        let start = min(Calendar.current.startOfDay(for: Date()), Date(timeIntervalSinceNow: -2 * loopManager.carbStore.defaultAbsorptionTimes.slow))
```
Comment (mpangburn) [inline WatchApp Extension/Controllers/CarbEntryListController.swift]: Clearer to use `-carbStore.maximumAbsorptionTimeInterval` over `-2 * carbStore.defaultAbsorptionTimes.slow`.

**Review by mpangburn — APPROVED** [review]
Just some reusability and clarity suggestions.
LGTM!

File: WatchApp Extension/Controllers/ChartHUDController.swift [hunk 22]
```diff
@@ -15,10 +15,32 @@ import SpriteKit
...
+        case reservoirVolume
+
+        var title: String {
+            switch self {
+            case .iob:
+                return NSLocalizedString("Active Insulin", comment: "HUD row title for IOB")
+            case .cob:
+                return NSLocalizedString("Active Carbs", comment: "HUD row title for COB")
+            case .netBasal:
+                return NSLocalizedString("Net Basal Rate", comment: "HUD row title")
```
Comment (darinkrauss) [inline WatchApp Extension/Controllers/ChartHUDController.swift]: This is the only comment that doesn't say what the "HUD row title" is for.

File: WatchApp Extension/Controllers/CarbEntryListController.swift [hunk 23]
```diff
@@ -0,0 +1,120 @@
...
+
+    private lazy var carbFormatter: QuantityFormatter = {
+        let formatter = QuantityFormatter()
+        formatter.numberFormatter.numberStyle = .none
+        return formatter
+    }()
+
+    private var observers: [Any] = [] {
+        didSet {
+            for observer in oldValue {
```
Comment (darinkrauss) [inline WatchApp Extension/Controllers/CarbEntryListController.swift:35]: What happens if the old observers and new observers overlap? I'm guessing this won't happen by the caller.

File: WatchApp Extension/Controllers/CarbEntryListController.swift [hunk 24]
```diff
@@ -0,0 +1,120 @@
...
+                    self.setCarbEntries(entries)
+                }
+            case .failure(let error):
+                self.log.error("Failed to fetch carb entries: %{public}@", String(describing: error))
+            }
+        }
+    }
+
+    private func setCarbEntries(_ entries: [StoredCarbEntry]) {
+        dispatchPrecondition(condition: .onQueue(.main))
```
Comment (darinkrauss) [inline WatchApp Extension/Controllers/CarbEntryListController.swift:95]: We don't do this in other places where we need this same precondition.

File: Loop.xcodeproj/project.pbxproj [hunk 25]
```diff
@@ -2636,7 +3160,7 @@
 				CLANG_WARN__DUPLICATE_METHOD_MATCH = YES;
 				CODE_SIGN_IDENTITY = "iPhone Developer: loudnate@gmail.com (XZN842LDLT)";
 				COPY_PHASE_STRIP = NO;
-				CURRENT_PROJECT_VERSION = 56;
+				CURRENT_PROJECT_VERSION = 55;
```
Comment (darinkrauss) [inline Loop.xcodeproj/project.pbxproj]: Did the project version go down?

File: Learn/Lessons/TimeInRangeLesson.swift [hunk 26]
```diff
@@ -0,0 +1,265 @@
...
+
+        self.configurationSections = [
+            LessonSection(headerTitle: nil, footerTitle: nil, cells: [dateEntry, weeksEntry]),
+            rangeEntry
+        ]
+    }
+
+    func execute(completion: @escaping ([LessonSectionProviding]) -> Void) {
+        guard let weeks = weeksEntry.number?.intValue, let closedRange = rangeEntry.closedRange else {
+            // TODO: Cleaner error presentation
```
Comment (darinkrauss) [inline Learn/Lessons/TimeInRangeLesson.swift:69]: TODO?

**Review by darinkrauss — APPROVED** [review]
LGTM 🚀 

There are few comments that might need some changes.

File: Learn/Extensions/NSNumber.swift (hunk 12 above)
Comment (ps2) [reply Learn/Extensions/NSNumber.swift]: Providing another initializer for optionals, one that returns nil if a nil was passed in seems straightforward and obvious to me at least.  Using map for a singular value, on the other hand, feels contorted.

Comment (mpangburn) [reply Learn/Extensions/NSNumber.swift]: The same argument could be made for any initializer, or more generally, any function: "apply this operation on the operands, but only if they're non-nil; if they're nil, just return nil".
More abstractly, this describes lifting a function of type `(A) -> B` into a function of type `(A?) -> B?` that "fails" if its argument is nil. That's exactly the type signature and purpose of `map` on Optional, and in fact very closely related to Swift's optional chaining operator.

If you feel strongly about this, the initializer is fine with me—but the definition of `map` on Optional is no contorsion. It's exactly the same level of abstraction as `map` on Array.
Happy to go into more detail here if it helps; this is a fascinating subject rooted in intersection of category theory and functional programming.

Comment (ps2) [reply Learn/Extensions/NSNumber.swift]: You kids get off my lawn!  

Ok, I did some reading on map & flatMap on optionals. I tend to prefer code that reads more clearly, rather than "clever" code. I'd have less of an issue with this if the practice of using map on optionals was more widespread.  Knowing how it works, I'm less opposed to it, but it does still feel like a bit of an obscurity.

Comment (ps2) [reply Learn/Extensions/NSNumber.swift]: And, after thinking about this more, I do like that it solves a not uncommon problem in a generic way.  The name is the thing that most throws me off.  Something like value.optionallyApply(myfunc) seems like it would be more obvious.  

Additionally, seeing a lot of this kind of thing might be a code smell, in that our code generally shouldn't have a lot of optional values lying around.

File: Learn/Lessons/TimeInRangeLesson.swift (hunk 17 above)
Comment (ps2) [reply Learn/Lessons/TimeInRangeLesson.swift]: Yeah, this seems highly reusable. Thanks!

File: Learn/View Controllers/LessonConfigurationViewController.swift (hunk 18 above)
Comment (ps2) [reply Learn/View Controllers/LessonConfigurationViewController.swift:161]: Yeah, it's a bit confusing. I'll see about getting this change into LoopKit

File: Learn/Extensions/NSNumber.swift (hunk 12 above)
Comment (mpangburn) [reply Learn/Extensions/NSNumber.swift]: I agree the name is less intuitive for Optional and hinders discoverability.

The benefit with using the term of art across types is that the generic abstraction becomes apparent:
`map` on Optional: `(A?, (A) -> B) -> B?`
`map` on Array: &nbsp; &nbsp; &nbsp;`([A], (A) -> B) -> [B]` 
`map` generically:  &nbsp;`(F<A>, (A) -> B) -> F<B>`

The overrides PR defines [`map` on DailyValueSchedule](https://github.com/tidepool-org/LoopKit/blob/c8e7dde4b659a3eb81ba5dc2e360621895e62f19/LoopKit/DailyValueSchedule.swift#L169):
`(DailyValueSchedule<A>, (A) -> B) -> DailyValueSchedule<B>`

The shared name allows someone familiar with concept of `map` in other contexts to see `map` on a new type and know what the function does without additional documentation. `flatMap` describes a related pattern where one layer of the structure is "flattened" when applying the transformation.

With Optional in particular, these pattern crop up often because UIKit classes often vend properties with Optional types for unclear reasons, like UITextField's `text: String?`, which doesn't have an obvious semantic distinction between `nil` and `""`. The optionality of these values then percolates down when applying transformations, which is unfortunate.

File: Loop/de.lproj/InfoPlist.strings (hunk 1 above)
Comment (ps2) [reply Loop/de.lproj/InfoPlist.strings:11]: Discussed in slack. We'll need to treat translations as unreviewed for now, and create a separate process for reviewing translations.

File: WatchApp Extension/Models/GlucoseChartData.swift (hunk 2 above)
Comment (ps2) [reply WatchApp Extension/Models/GlucoseChartData.swift:98]: We should not offload these kinds of decisions onto the user if we can come up with values that work for most.  I'll bring the mmol/L number more in line with the mg/dl version.

File: WatchApp/Base.lproj/Interface.storyboard (hunk 3 above)
Comment (ps2) [reply WatchApp/Base.lproj/Interface.storyboard:28]: Yes, the colors are intentionally now opaque. Also, they’re computed as opaque for black background, not white.

File: WatchApp/Base.lproj/Interface.storyboard (hunk 5 above)
Comment (ps2) [reply WatchApp/Base.lproj/Interface.storyboard:546]: I believe this is just xcode cache.

File: WatchApp Extension/Scenes/GlucoseChartValueHashable.swift (hunk 6 above)
Comment (ps2) [reply WatchApp Extension/Scenes/GlucoseChartValueHashable.swift:81]: GlucoseChartValueHashable min and max properties are always milligramsPerDeciliter

File: WatchApp Extension/Managers/LoopDataManager.swift (hunk 7 above)
Comment (ps2) [reply WatchApp Extension/Managers/LoopDataManager.swift:66]: The most you can see on the graph is around 3 hours.

File: LoopCore/CarbEntryUserInfo.swift (hunk 8 above)
Comment (ps2) [reply LoopCore/CarbEntryUserInfo.swift:29]: This is used when sending a carb entry from the watch to the phone; since the watch app is usually updated when the app is, we shouldn't generally have a mismatch.

File: Loop/View Controllers/ChartsTableViewController.swift (hunk 10 above)
Comment (ps2) [reply Loop/View Controllers/ChartsTableViewController.swift:175]: reloadData() has a lot of assertions about various states, so this would not actually optimize anything.  Also, if we needed to pause something, it would probably make more sense to do it in reloadData() rather than putting in the didSet block.

File: LoopUI/Managers/StatusChartsManager.swift (hunk 13 above)
Comment (ps2) [reply LoopUI/Managers/StatusChartsManager.swift:956]: If there isn't an override, the pointsForGlucoseRangeScheduleOverride returns [], so the conditional block was unnecessary.

File: WatchApp Extension/Controllers/CarbEntryListController.swift (hunk 23 above)
Comment (ps2) [reply WatchApp Extension/Controllers/CarbEntryListController.swift:35]: addObserver returns a unique opaque object for each subscription, so they can't overlap.

File: WatchApp Extension/Controllers/CarbEntryListController.swift (hunk 24 above)
Comment (ps2) [reply WatchApp Extension/Controllers/CarbEntryListController.swift:95]: data model updates are probably the most likely functions to be called from other queues.  This is defensive programming, and as such should only be used judiciously.

File: Loop.xcodeproj/project.pbxproj (hunk 25 above)
Comment (ps2) [reply Loop.xcodeproj/project.pbxproj]: This is due to a wrinkle in the space-time continuum. I'll bump it.

File: Learn/Lessons/TimeInRangeLesson.swift (hunk 26 above)
Comment (ps2) [reply Learn/Lessons/TimeInRangeLesson.swift:69]: Yeah, this probably deserves a larger thinking out of input validation, or how to make the inputs not allow erroneous input.

**Comment by mpangburn** [comment]
LGTM, no additional comments.

**Review by darinkrauss — APPROVED** [review]
LGTM 🚀
